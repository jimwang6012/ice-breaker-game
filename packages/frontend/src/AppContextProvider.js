import React from "react";
import { useListState } from "@mantine/hooks";
import { useState } from "react";
import useAudio from "./hook/useAudio";

export const AppContext = React.createContext({
  players: [],
  handlers: null,
  isHost: false,
  setIsHost: (isHost) => {},
  roomId: "",
  setRoomId: (roomId) => {},
  hostId: "",
  setHostId: (hostId) => {},
  name: "",
  setName: (name) => {},
  isTyping: false,
  setIsTyping: (isTyping) => {},
  config: {
    roomSize: 8,
    boardSize: 9,
    roundTime: 30,
    breakTime: 0.5,
  },
  setConfig: (config) => {},
  colors: [],
  setColors: (colors) => {},
  colorStatus: [],
  setColorStatus: (colorStatus) => {},
  playing: true,
  toggle: () => {},
  stop: () => {},
});

export function AppContextProvider({ children }) {
  // Manage players in the room
  const [players, handlers] = useListState();

  // Whether the player is a host
  const [isHost, setIsHost] = useState(false);

  // The current room id
  const [roomId, setRoomId] = useState("");

  // The host's id
  const [hostId, setHostId] = useState("");

  //player name
  const [name, setName] = useState("");

  //game config
  const [config, setConfig] = useState({
    roomSize: 8,
    boardSize: 9,
    roundTime: 30,
    breakTime: 0.5,
  });

  //player colors
  const [colors, setColors] = useState([]);

  //player color status
  const [colorStatus, setColorStatus] = useState([]);

  //isChatting
  const [isTyping, setIsTyping] = useState(false);

  const { playing, toggle, stop } = useAudio("../music/music2.ogg");

  const context = {
    players,
    handlers,
    isHost,
    setIsHost,
    roomId,
    setRoomId,
    hostId,
    setHostId,
    name,
    setName,
    config,
    setConfig,
    colors,
    setColors,
    colorStatus,
    setColorStatus,
    isTyping,
    setIsTyping,
    playing,
    toggle,
    stop,
  };

  return <AppContext.Provider value={context}>{children}</AppContext.Provider>;
}
