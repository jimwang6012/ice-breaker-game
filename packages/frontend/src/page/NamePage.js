import React, { useContext, useState } from "react";
import { useNavigate } from "react-router-dom";
import { AppContext } from "../AppContextProvider";
import socket from "../Socket";
import { MainButton } from "../component/MainButton";
import { Modal } from "../component/Modal";
import { BsFillXCircleFill, BsCheckCircleFill } from "react-icons/bs";

function NamePage() {
  document.body.style.overflow = "hidden";
  const navigate = useNavigate();
  const {
    isHost,
    setRoomId,
    roomId,
    handlers,
    name,
    setName,
    setConfig,
    setColors,
    setColorStatus,
  } = useContext(AppContext);

  const toHome = () => {
    navigate("/home");
  };

  const [roomPrmpt, setmessgae] = useState("");
  const [roomNotFound, setNotFound] = useState(false);
  const [show, setShow] = useState(false);
  const createRoom = () => {
    socket.connect();
    socket.emit("create-room", { name: name }, (room) => {
      const string = "Room " + room.roomId + " Created!";
      setmessgae(string);
      setShow(true);
      // alert("Room " + room.roomId + " Created by " + name);
      setRoomId(room.roomId);
      handlers.setState(room.players);
      setName(name);
      setConfig(room.config);
      setColors(room.colors);
      setColorStatus(room.colorStatus);
      navigate("/" + room.roomId.toString() + "/idle");
    });
  };

  const joinRoom = () => {
    socket.connect();
    socket.emit("join-room", { roomId, name }, (room) => {
      if (room) {
        setRoomId(room.roomId);
        handlers.setState(room.players);
        setName(name);
        setConfig(room.config);
        setColors(room.colors);
        setColorStatus(room.colorStatus);
        navigate("/" + roomId.toString() + "/idle");
      } else {
        setmessgae("Room Not Found!");
        setNotFound(true);
        setShow(true);
      }
    });
  };

  return (
    <div className="overflow-hidden">
      <Modal
        show={show}
        pageJump={() => {
          if (!roomNotFound) {
            navigate("/" + roomId + "/idle");
          } else {
            navigate("/home");
          }
        }}
        mainPrompt={roomPrmpt}
        buttonPrompt={!roomNotFound ? "Join" : "OK"}
        title={
          <div className="pb-3 flex flex-row justify-center text-4xl text-ice-7">
            {!roomNotFound ? <BsCheckCircleFill /> : <BsFillXCircleFill />}
          </div>
        }
      />
      <div className="flex items-center justify-center h-screen overflow-y-auto bg-ice-8 opacity-90">
        {/* left square */}
        <div className="absolute top-0 rotate-45 -left-96 h-2/3 aspect-square bg-ice-3" />
        <div className="z-40 flex flex-col items-center justify-center gap-20 ">
          <div className="text-6xl font-bold text-white">Ice Breakers!</div>
          {/* center modal */}
          <div className="py-16 rounded-lg shadow-lg px-28 bg-ice-2">
            <div className="flex flex-row justify-start gap-10">
              <input
                className="justify-center text-xl font-bold text-center border rounded-md text-ice-7 px-14 bg-ice-0 border-ice-6"
                placeholder="Enter your name"
                aria-label="dwe"
                maxLength={12}
                onChange={(e) => {
                  setName(e.target.value);
                }}
              />
              {isHost ? (
                <MainButton handleClick={createRoom} text="Confirm" />
              ) : (
                <MainButton handleClick={joinRoom} text="Confirm" />
              )}
            </div>
          </div>
          <MainButton handleClick={toHome} text="Back" />
        </div>
        {/* right square */}
        <div className="absolute bottom-0 rotate-45 -right-96 h-2/3 aspect-square bg-ice-3" />
      </div>
    </div>
  );
}

export default NamePage;
