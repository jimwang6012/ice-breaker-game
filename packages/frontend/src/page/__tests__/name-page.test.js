import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import { AppContextProvider } from "../../AppContextProvider";
import NamePage from "../NamePage";

it("renders correctly", () => {
  const tree = renderer
    .create(
      <AppContextProvider>
        <BrowserRouter>
          <NamePage />
        </BrowserRouter>
      </AppContextProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
