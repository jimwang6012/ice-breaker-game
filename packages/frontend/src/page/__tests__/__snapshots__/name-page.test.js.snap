// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="overflow-hidden"
>
  <div
    className="modal display-none"
  >
    <section
      className="modal-main"
    >
      <div
        className=""
      >
        <div
          className="p-5 rounded-lg shadow dark:bg-ice-8"
        >
          <div
            className="text-center "
          >
            <div
              className="pb-3 flex flex-row justify-center text-4xl text-ice-7"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 16 16"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M16 8A8 8 0 1 1 0 8a8 8 0 0 1 16 0zm-3.97-3.03a.75.75 0 0 0-1.08.022L7.477 9.417 5.384 7.323a.75.75 0 0 0-1.06 1.06L6.97 11.03a.75.75 0 0 0 1.079-.02l3.992-4.99a.75.75 0 0 0-.01-1.05z"
                />
              </svg>
            </div>
            <h3
              className="mb-5 text-xl font-bold text-ice-7"
            >
              
            </h3>
            <button
              className="px-5 py-3 mt-5 font-semibold text-white rounded-lg text-l bg-ice-6 hover:bg-ice-5"
              onClick={[Function]}
            >
              Join
            </button>
          </div>
        </div>
      </div>
    </section>
  </div>
  <div
    className="flex items-center justify-center h-screen overflow-y-auto bg-ice-8 opacity-90"
  >
    <div
      className="absolute top-0 rotate-45 -left-96 h-2/3 aspect-square bg-ice-3"
    />
    <div
      className="z-40 flex flex-col items-center justify-center gap-20 "
    >
      <div
        className="text-6xl font-bold text-white"
      >
        Ice Breakers!
      </div>
      <div
        className="py-16 rounded-lg shadow-lg px-28 bg-ice-2"
      >
        <div
          className="flex flex-row justify-start gap-10"
        >
          <input
            aria-label="dwe"
            className="justify-center text-xl font-bold text-center border rounded-md text-ice-7 px-14 bg-ice-0 border-ice-6"
            maxLength={12}
            onChange={[Function]}
            placeholder="Enter your name"
          />
          <div
            className="flex"
          >
            <button
              className="absolute z-40 px-6 py-3 text-xl font-semibold text-white rounded-lg shadow-inner bg-ice-6 hover:mt-1 active:mt-2"
              onClick={[Function]}
            >
              Confirm
            </button>
            <div
              className="px-6 py-3 mt-2 text-xl font-semibold rounded-lg text-ice-7 bg-ice-7"
            >
              Confirm
            </div>
          </div>
        </div>
      </div>
      <div
        className="flex"
      >
        <button
          className="absolute z-40 px-6 py-3 text-xl font-semibold text-white rounded-lg shadow-inner bg-ice-6 hover:mt-1 active:mt-2"
          onClick={[Function]}
        >
          Back
        </button>
        <div
          className="px-6 py-3 mt-2 text-xl font-semibold rounded-lg text-ice-7 bg-ice-7"
        >
          Back
        </div>
      </div>
    </div>
    <div
      className="absolute bottom-0 rotate-45 -right-96 h-2/3 aspect-square bg-ice-3"
    />
  </div>
</div>
`;
